spopt/region/sa3.py

```diff
@@ -0,0 +1,221 @@
+
+
+def extract_clusters(linkage_matrix, min_cluster_size, eom_clusters=True):
+    """Extract cluster types from a linkage matrix.
+
+    Parameters
+    ----------
+
+    linkage_matrix : np.ndarray
+        A hierarchical clustering encoded as an array in
+        ``scipy.cluster.hierarchy.linkage`` format.
+    min_cluster_size : int
+        The minimum number of observations that forms a cluster.
+    eom_clusters: boolean (default True)
+        The cluster extraction scheme. True is Excess of Mass, False - Leaf.
+
+    Returns
+    -------
+    numpy.array
+        Cluster labels for observations.
+
+    Note
+    ----
+    This function requires ``fast_hdbscan`` and ``numba``.
+
+    """
+
+    n_samples = linkage_matrix.shape[0] + 1
+
+    try:
+        from fast_hdbscan.cluster_trees import (
+            cluster_tree_from_condensed_tree,
+            condense_tree,
+            extract_eom_clusters,
+            extract_leaves,
+            get_cluster_label_vector,
+        )
+    except ImportError as e:
+        raise ImportError(
+            "The fast_hdbscan and numba libraries are required for this functionality."
+        ) from e
+
+    condensed_tree = condense_tree(linkage_matrix, min_cluster_size=min_cluster_size)
+
+    cluster_tree = cluster_tree_from_condensed_tree(condensed_tree)
+
+    if eom_clusters:
+        selected_clusters = extract_eom_clusters(
+            condensed_tree, cluster_tree, allow_single_cluster=False
+        )
+    else:
+        selected_clusters = extract_leaves(condensed_tree, allow_single_cluster=False)
+
+    return get_cluster_label_vector(condensed_tree, selected_clusters, 0, n_samples)
+
+
+class SA3(BaseSpOptHeuristicSolver):
+    """Spatial Adaptive Agglomerative Aggregation (SA3) clustering.
+
+    The algorithm carries out ``sklearn.cluster.AgglometariveClustering``
+    per the specified parameters and extracts clusters from it, using density-clustering
+    extraction algorithms - Excess of Mass or Leaf. This results in multiscale,
+    contiguous clusters with noise.
+
+
+
+    Parameters
+    ----------
+
+    gdf : geopandas.GeoDataFrame
+        Geodataframe containing original data.
+    w : libpysal.weights.W | libpysal.graph.Graph
+        Weights or Graph object created from given data.
+    attrs_name : list
+        Strings for attribute names (cols of ``geopandas.GeoDataFrame``).
+    min_cluster_size : int
+        The minimum number of observations to form a cluster.
+    eom_clusters: boolean (default True)
+        The cluster extraction scheme. True is Excess of Mass, False - Leaf.
+    clustering_kwds: dict
+        Parameters about clustering could be used in
+        ``sklearn.cluster.AgglometariveClustering.``
+
+    Attributes
+    ----------
+
+    labels_ : numpy.array
+        Cluster labels for observations.
+
+    """
+
+    def __init__(
+        self,
+        gdf,
+        w,
+        attrs_name,
+        min_cluster_size=15,
```

Comment:
The default is meaningless without knowing the use case. Shall we maybe make it a required arg?